presto-cassandra/src/main/java/com/facebook/presto/cassandra/CassandraClusteringPredicatesExtractor.java

```diff
@@ -87,75 +63,139 @@ public CassandraClusteringPredicatesExtractor(List<CassandraColumnHandle> cluste
+                                            columnHandle.getCassandraType()));
+                                    break;
+                                case BELOW:
+                                    rangeConjuncts.add(CassandraCqlUtils.validColumnName(columnHandle.getName()) + " < "
+                                            + CassandraCqlUtils.cqlValue(toCQLCompatibleString(range.getHigh().getValue()),
+                                            columnHandle.getCassandraType()));
+                                    break;
+                                default:
+                                    throw new AssertionError("Unhandled bound: " + range.getHigh().getBound());
+                            }
                         }
-                        return ImmutableSet.of();
-                    },
-                    allOrNone -> ImmutableSet.of());
+                    }
+                }
 
-            if (!values.isEmpty()) {
-                clusteringColumnValues.add(values);
+                if (!singleValues.isEmpty()) {
+                    if (singleValues.size() == 1) {
+                        predicate = CassandraCqlUtils.validColumnName(columnHandle.getName()) + " = " + singleValues.get(0);
+                    }
+                    else {
+                        predicate = CassandraCqlUtils.validColumnName(columnHandle.getName()) + " IN ("
+                                + Joiner.on(",").join(singleValues) + ")";
+                    }
+                }
+                else if (!rangeConjuncts.isEmpty()) {
+                    predicate = Joiner.on(" AND ").join(rangeConjuncts);
+                }
+                return predicate;
+            }, discreteValues -> {
+                if (discreteValues.isWhiteList()) {
+                    ImmutableList.Builder<Object> discreteValuesList = ImmutableList.builder();
+                    for (Object discreteValue : discreteValues.getValues()) {
+                        discreteValuesList.add(CassandraCqlUtils.cqlValue(toCQLCompatibleString(discreteValue),
+                                columnHandle.getCassandraType()));
+                    }
+                    String predicate = CassandraCqlUtils.validColumnName(columnHandle.getName()) + " IN ("
+                            + Joiner.on(",").join(discreteValuesList.build()) + ")";
+                    return predicate;
+                }
+                return null;
+            }, allOrNone -> null);
+
+            if (predicateString != null) {
+                // Range all not supported
+                if (predicateString.equals("Not Supported")) {
+                    break;
+                }
+                // IN restriction only on last clustering column for Cassandra version < 2.2
+                if (predicateString.contains(" IN (") && cassandraVersion != null && Float.parseFloat(cassandraVersion.substring(0, 3)) < 2.2 && currentClusteringColumn++ != (clusteringColumns.size() - 1)) {
```

Comment:
I have added a new class `CassandraVersion`. Instead of `regex`, I have converted this to `Flot `to do the inequality check. Kindly review once and let me know how we can improve this.

Also for version > 2.1, `IN `clause can be anywhere in CQL string.
and for version = 2.1, `IN `clause has to be part of only last clustering column of a table, can't have any other place even if it is applied to last column being restricted which is not the last clustering column.

`cqlsh:smoke_test> _select * from table_clustering_keys;_

 key | clust_one | clust_two | clust_three | data
-----+-----------+-----------+-------------+------

(0 rows)
cqlsh:smoke_test> _select * from table_clustering_keys where key = '1' and clust_one IN ('a','b');_
**InvalidRequest: code=2200 [Invalid query] message="Clustering column "clust_one" cannot be restricted by an IN relation"**
cqlsh:smoke_test> _select * from table_clustering_keys where key = '1' and clust_one = 'a' and clust_two = 'b' and clust_three = 'c'_;

 key | clust_one | clust_two | clust_three | data
-----+-----------+-----------+-------------+------

(0 rows)
cqlsh:smoke_test> _select * from table_clustering_keys where key = '1' and clust_one = 'a' and clust_two = 'b' and clust_three IN ('a','c');_

 key | clust_one | clust_two | clust_three | data
-----+-----------+-----------+-------------+------

(0 rows)
cqlsh:smoke_test> show VERSION ;
[cqlsh 5.0.1 | Cassandra 2.1.5 | CQL spec 3.2.0 | Native protocol v3]
cqlsh:smoke_test>
`